selftests/smbus_test.c

```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: GPL-2.0
+//
+// SPD files can be also pened with:
+//   decode-dimms -x <(xxd data/<filename.spd>)
+
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "smbus.h"
+#include "selftests.h"
+
+// Make the linker happy.
+struct mem_dev *dmi_memory_device;
+void *pci_config_read16;
+void *pci_config_read8;
+void *pci_config_write8;
+void *imc_type;
+void *quirk;
+
+#define MAX_SPD_SIZE 1024
+
+#define parse_spd_ddr4 parse_spd_ddr4_wrapper
+void parse_spd_ddr4(spd_info *spdi, uint8_t slot_idx);
+
+#define parse_spd_ddr3 parse_spd_ddr3_wrapper
+void parse_spd_ddr3(spd_info *spdi, uint8_t slot_idx);
+
+#define parse_spd_ddr2 parse_spd_ddr2_wrapper
+void parse_spd_ddr2(spd_info *spdi, uint8_t slot_idx);
+
+#define parse_spd_ddr parse_spd_ddr_wrapper
+void parse_spd_ddr(spd_info *spdi, uint8_t slot_idx);
+
+#define print_spdi print_spdi_wrapper
+void print_spdi(spd_info spdi, uint8_t lidx);
+
+uint8_t spd[MAX_SPD_SIZE];
+uint16_t spd_size;
+uint8_t spd_slot;
+
+void load_spd(uint8_t slot_idx, char *filename) {
+    int f = open(filename, 0);
+
+    if (f == -1) {
+        fprintf(stderr, "Cannot open %s\n", filename);
+        exit(1);
+    }
+
+    spd_size = read(f, spd, MAX_SPD_SIZE);
+    close(f);
+
+    spd_slot = slot_idx;
+    printf("Loaded %s, slot=%u, size=%u\n", filename, slot_idx, spd_size);
+}
+
+uint8_t mock_read_spd_byte(uint8_t slot_idx, uint16_t spd_adr) {
+
+    if (spd_adr < spd_size && slot_idx == spd_slot)
+        return spd[spd_adr];
+
+    fprintf(stderr, "Unbound SPD read slot %u, addr %u\n", spd_slot, spd_adr);
+    exit(1);
+}
+
+void test_parse_spd_ddr4() {
+
+    spd_info spdi;
+    uint8_t slot_idx = 0;
+
+    // BUG: shuld be reported as DDR4-2133 15-15-15-36 not DDR4-2132 14-14-14-35
```

Comment:
Typo: "shuld" (here and below)

Reply:
Ah, yes, thanks. Luckily, they are no longer needed as the SPD parsing now works as expected for every testcase I have. :)